### PR TITLE
privilege,backupccl: introduce a RESTORE privilege

### PR DESCRIPTION
**Pull request by adityamaru**
This change introduces a RESTORE privilege that is grantable
as a system, or database level privilege. The purpose
of this privilege is to offer a fine-grained permission model
for users that wish to run restores. First let us outline the
existing privilege model that governs backups:

Cluster restores - admin only.

Database restores - users must have CREATEDB role option.

Table restores - users must have CREATE on the database we
are restoring into.

With the new fine grained permission model we would like to stop
having to grant users privileges such as CREATE and CREATEDB only
for the purpose of being able to restore a target. To this effect
the new privilege model that will govern restores is:

Cluster backups - user requires the system RESTORE privilege

Database backups - user requires the system RESTORE privilege

Table backups - user requires the database RESTORE privilege

Note, admins will ofcourse continue to bypass all these checks.
This diff does not change the privilege checks we perform on the backup
destination URIs related to IMPLICIT authentication. That will be done
as a follow-up.

In 22.2 to prevent breaking user flows we will continue to respect the
old privilege model, but emit a notice indicating our plans to replace
this model in 23.1. At which point users will need to be granted the
appropriate BACKUP privileges.

Informs: https://github.com/cockroachdb/cockroach/issues/86263

Release note (sql change): This change introduces a new RESTORE
privilege that is grantable as a system or database level privilege.
Users can opt-in to the new privilege model by granting the appropriate
privileges as per the following model:

Cluster backups - user requires the system RESTORE privilege

Database backups - user requires the system RESTORE privilege

Table backups - user requires the database RESTORE privilege

In 22.2 we will continue to respect the old privilege model, but will
completely swithover to the RESTORE privilege in 23.1.

Release justification: high impact change to offer fine grained privileges
for bulk operations

### CODE REVIEW

**Comment by adityamaru** [comment]
Only the last commit is new.

**Review by stevendanna — COMMENTED** [review]
Overall last commit looks good. Only blocking thing is a small typo in one of the error messages.

File: pkg/ccl/backupccl/restore_planning.go [hunk 1]
```diff
@@ -653,12 +698,12 @@ func getDatabaseIDAndDesc(
 func dropDefaultUserDBs(ctx context.Context, execCfg *sql.ExecutorConfig) error {
 	return sql.DescsTxn(ctx, execCfg, func(ctx context.Context, txn *kv.Txn, col *descs.Collection) error {
 		ie := execCfg.InternalExecutor
-		_, err := ie.Exec(ctx, "drop-defaultdb", nil, "DROP DATABASE IF EXISTS defaultdb")
+		_, err := ie.Exec(ctx, "drop-defaultdb", txn, "DROP DATABASE IF EXISTS defaultdb")
```
[the remark this answers is not in the document]
Comment (adityamaru) [reply pkg/ccl/backupccl/restore_planning.go:701]: yeah...going to pull this one out maybe or backport it for sure. Possibly related thread - https://cockroachlabs.slack.com/archives/C2C5FKPPB/p1661202884691609

File: pkg/ccl/backupccl/restore_planning.go [hunk 2]
```diff
@@ -137,6 +137,28 @@ func maybeFilterMissingViews(
...
+func checkRestorePrivilegesOnDatabase(
+	ctx context.Context, p sql.PlanHookState, parentDB catalog.DatabaseDescriptor,
+) (shouldBufferNotice bool, err error) {
+	if err := p.CheckPrivilege(ctx, parentDB, privilege.RESTORE); err == nil {
+		return false, nil
+	}
+
+	if err := p.CheckPrivilege(ctx, parentDB, privilege.CREATE); err != nil {
+		preamble := "The existing privileges required for restore are being deprecated " +
+			"in favour of a fine-grained privilege model explained here <link>. In a future release, to run"
```
[the remark this answers is not in the document]
Comment (adityamaru) [reply pkg/ccl/backupccl/restore_planning.go]: done!

**Comment by adityamaru** [comment]
Flake is `TestCCLLogic_multi_region_remote_access_error` which is being discussed in sql-experience. @stevendanna I believe I need a stamp before I can bors, thanks!

**Review by stevendanna — APPROVED** [review]

**Comment by adityamaru** [comment]
TFTR!

bors r=stevendanna

**Pull request merged**